frontend/dashboard/widget/controller.py

```diff
@@ -0,0 +1,379 @@
+        self._last_state = state
+
+    def update_packet_count(self, state):
+        #Keep track of received and missed packets for performance tuning
+        self.received_packets += 1
+        missed_packets = state.packet_number - \
+                         self._last_state.packet_number - 1
+        if missed_packets:
+            self.dispatch_event('on_missed_packet', missed_packets)
+        self.missed_packets += missed_packets
+
+    def handle_changed_state(self, state):
+        "Dispatch various events as a result of the state changing"
+        self.dispatch_event('on_state_changed', state)
+        self.dispatch_axis_events(state)
+        self.dispatch_button_events(state)
+
+    def dispatch_axis_events(self, state):
+        # axis fields are everything but the buttons
+        axis_fields = dict(XINPUT_GAMEPAD._fields_)
+        axis_fields.pop('buttons')
+        for axis, type in list(axis_fields.items()):
+            old_val = getattr(self._last_state.gamepad, axis)
+            new_val = getattr(state.gamepad, axis)
+            data_size = ctypes.sizeof(type)
+            old_val = self.translate(old_val, data_size)
+            new_val = self.translate(new_val, data_size)
+
+            # an attempt to add deadzones and dampen noise
+            # done by feel rather than following http://msdn.microsoft.com/en-gb/library/windows/desktop/ee417001%28v=vs.85%29.aspx#dead_zone
+            # ags, 2014-07-01
+            if ((old_val != new_val and (new_val > 0.08000000000000000 or new_val < -0.08000000000000000) and abs(
+                    old_val - new_val) > 0.00000000500000000) or
+                    (axis == 'right_trigger' or axis == 'left_trigger') and new_val == 0 and abs(
+                        old_val - new_val) > 0.00000000500000000):
+                self.dispatch_event('on_axis', axis, new_val)
+
+    def dispatch_button_events(self, state):
+        changed = state.gamepad.buttons ^ self._last_state.gamepad.buttons
+        changed = get_bit_values(changed, 16)
+        buttons_state = get_bit_values(state.gamepad.buttons, 16)
+        changed.reverse()
+        buttons_state.reverse()
+        button_numbers = count(1)
+        changed_buttons = list(
+            filter(itemgetter(0), list(zip(changed, button_numbers, buttons_state))))
+        tuple(starmap(self.dispatch_button_event, changed_buttons))
+
+    def dispatch_button_event(self, changed, number, pressed):
+        self.dispatch_event('on_button', number, pressed)
+
+    # stub methods for event handlers
+    def on_state_changed(self, state):
+        pass
+
+    def on_axis(self, axis, value):
+        pass
+
+    def on_button(self, button, pressed):
+        pass
+
+    def on_missed_packet(self, number):
+        pass
+#Definitions for the joystick methods
+list(map(XInputJoystick.register_event_type, [
+    'on_state_changed',
+    'on_axis',
+    'on_button',
+    'on_missed_packet',
+]))
```

Comment:
The widget should only have stuff to update an image of a controller to show what buttons are being pressed or something. 

Can you move all controller code into `p2s/manipulator/XInputJoystick.py` and `p2s/manipulator/manipulator.py`.

Reply:
XInput looks tinteresting. 

@davidxiang101 is also working on controller code rn, so you can talk to him on slack about it.

Reply:
lol it works